sdk/program/src/lib.rs

```diff
@@ -1,3 +1,650 @@
+//!         pubkey::Pubkey,
+//!     };
+//!
+//!     entrypoint!(process_instruction);
+//!
+//!     pub fn process_instruction(
+//!         program_id: &Pubkey,
+//!         accounts: &[AccountInfo],
+//!         instruction_data: &[u8],
+//!     ) -> ProgramResult {
+//!         // Decode and dispatch instructions here.
+//!         todo!()
+//!     }
+//! }
+//!
+//! // Additional code goes here.
+//! ```
+//!
+//! With a `Cargo.toml` file that contains
+//!
+//! ```toml
+//! [lib]
+//! crate-type = ["cdylib", "rlib"]
+//!
+//! [features]
+//! no-entrypoint = []
+//! ```
+//!
+//! Note that a Solana program must specify its crate-type as "cdylib", and
+//! "cdylib" crates will automatically be discovered and built by the `cargo
+//! build-bpf` command. Solana programs also often have crate-type "rlib" so
+//! they can be linked to other Rust crates.
+//!
+//! # On-chain vs. off-chain compilation targets
+//!
+//! Solana programs run on the [rbpf] VM, which implements a variant of the
+//! [eBPF] instruction set. Because this crate can be compiled for both on-chain
+//! and off-chain execution, the environments of which are significantly
+//! different, it extensively uses [conditional compilation][cc] to tailor its
+//! implementation to the environment. The `cfg` predicate used for identifying
+//! compilation for on-chain programs is `target_arch = "bpf"`, as in this
+//! example from the `solana-program` codebase that logs a message via a
+//! syscall when run on-chain, and via a library call when offchain:
+//!
+//! [rbpf]: https://github.com/solana-labs/rbpf
+//! [eBPF]: https://ebpf.io/
+//! [cc]: https://doc.rust-lang.org/reference/conditional-compilation.html
+//!
+//! ```
+//! pub fn sol_log(message: &str) {
+//!     #[cfg(target_arch = "bpf")]
+//!     unsafe {
+//!         sol_log_(message.as_ptr(), message.len() as u64);
+//!     }
+//!
+//!     #[cfg(not(target_arch = "bpf"))]
+//!     program_stubs::sol_log(message);
+//! }
+//! # mod program_stubs {
+//! #     pub(crate) fn sol_log(message: &str) { }
+//! # }
+//! ```
+//!
+//! This `cfg` pattern is suitable as well for user code that needs to work both
+//! on-chain and off-chain.
+//!
+//! `solana-program` and `solana-sdk` were previously a single crate. Because of
+//! this history, and because of the dual-usage of `solana-program` for two
+//! different environments, it contains some features that are not available to
+//! on-chain programs at compile-time. It also contains some on-chain features
+//! that will fail in off-chain scenarios at runtime. This distinction is not
+//! well-reflected in the documentation.
+//!
+//! For a more complete description of Solana's implementation of eBPF and its
+//! limitations, see the main Solana documentation on [on-chain programs][ocp].
+//!
+//! [ocp]: https://docs.solana.com/developing/on-chain-programs/overview
+//!
+//! # Core data types
+//!
+//! - [`Pubkey`] - An [ed25519] public key. All Solana accounts are identified
```

Comment:
We often refer in the docs and elsewhere to account "addresses" and the `Pubkey` is how addresses are expressed.  Public key language is more commonly used when paired with a private key.  Often account addresses don't have private keys or the private keys aren't relevant to the use of the address.